### PR TITLE
Refactor nn.networks

### PR DESCRIPTION
**Pull request by sian-chen**
- Modularize common components.
- Add label-wise attention networks with BiGRU, BiLSTM, CNN encoders
- Refactor KimCNN and XMLCNN
- Remove BaseModel

### CODE REVIEW

File: libmultilabel/nn/networks/modules.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    """
+
+    def __init__(self, input_size, num_classes):
+        super(LabelwiseAttention, self).__init__()
+        self.attention = nn.Linear(input_size, num_classes, bias=False)
+
+    def forward(self, inputs):
+        attention = self.attention(inputs).transpose(1, 2)  # N, num_classes, L
+        attention = F.softmax(attention, -1)
+        return attention @ inputs   # N, num_classes, input_size
```
Comment (sian-chen) [inline libmultilabel/nn/networks/modules.py:160]:
```suggestion
        return attention @ inputs   # N, num_classes, input_size
```
Use bmm() function.

File: libmultilabel/nn/networks/caml.py [hunk 2]
```diff
@@ -2,12 +2,11 @@
...
+import torch.nn.functional as F
 from torch.nn.init import xavier_uniform_
 
-from ..networks.base import BaseModel
 
-
-class CAML(BaseModel):
+class CAML(nn.Module):
     """CAML (Convolutional Attention for Multi-Label classification)
     Follows the work of Mullenbach et al. [https://aclanthology.org/N18-1100.pdf]
```
Comment (sian-chen) [inline libmultilabel/nn/networks/caml.py:11]: Only for reproducibility.

File: libmultilabel/nn/networks/labelwise_attention_networks.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        hidden_dim = num_filter_per_size * len(filter_sizes)
+        super(CNNLWAN, self).__init__(embed_vecs, num_classes, dropout,
+                                      hidden_dim)
+
+    def _get_encoder(self, input_size, dropout):
+        return CNNEncoder(input_size, self.filter_sizes,
+                          self.num_filter_per_size, self.activation,
+                          channel_last=True)
+
+    def forward(self, input):
```
Comment (cjlin1) [inline libmultilabel/nn/networks/labelwise_attention_networks.py:134]: cannot we just use forward of LabelwiseAttentionNetwork?

File: libmultilabel/nn/networks/modules.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        return self.dropout(self.embedding(inputs))
+
+
+class RNNEncoder(ABC, nn.Module):
+    """Base class of RNN encoder with dropout
+
+    Args:
+        input_size (int): The number of expected features in the input.
+        hidden_size (int): The number of features in the hidden state.
+        num_layers (int): Number of recurrent layers.
```
Comment (cjlin1) [inline libmultilabel/nn/networks/modules.py:33]: "Number of .." different from "The number of" of the previous two arguments. Same issue in other class as well

File: libmultilabel/nn/networks/modules.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+class CNNEncoder(nn.Module):
+    """Multi-filter-size CNN encoder for text classification with max-pooling
+
+    Args:
+        input_size (int): The number of expected features in the input.
+        filter_sizes (list): Size of convolutional filters.
+        num_filter_per_size (int): Number of filters in convolutional layers in each size. Defaults to 128.
+        activation (str): Activation function to be used. Defaults to 'relu'.
+        num_pool (int): Number of pool for max-pooling.
```
Comment (cjlin1) [inline libmultilabel/nn/networks/modules.py:102]: pools

File: example_config/EUR-Lex-57k/bigru_lwan.yml [hunk 6]
```diff
@@ -20,12 +20,11 @@ monitor_metrics: ['Micro-F1', 'Macro-F1', 'Another-Macro-F1', 'P@1', 'P@5', 'RP@
 val_metric: P@1
 
 # model
-model_name: BiGRU
-init_weight: null
+model_name: BiGRULWAN
```
Comment (Eleven1Liu) [inline example_config/EUR-Lex-57k/bigru_lwan.yml:23]: Should  `example_config/MIMIC-50/bigru.yml` be updated?
https://github.com/ASUS-AICS/LibMultiLabel/blob/master/example_config/MIMIC-50/bigru.yml

File: libmultilabel/nn/networks/labelwise_attention_networks.py (hunk 3 above)
Comment (sian-chen) [reply libmultilabel/nn/networks/labelwise_attention_networks.py:134]: Because the RNN encoders require lengths but CNN is not.

File: libmultilabel/nn/networks/modules.py (hunk 4 above)
Comment (sian-chen) [reply libmultilabel/nn/networks/modules.py:33]: It seems we don't have a consistent rule for this kind of comment style yet.
Maybe discuss on Mon and revise all of them in the next PR?

File: example_config/EUR-Lex-57k/bigru_lwan.yml (hunk 6 above)
Comment (sian-chen) [reply example_config/EUR-Lex-57k/bigru_lwan.yml:23]: Updated, thanks.

**Review by cjlin1 — APPROVED** [review]

**Review by cjlin1 — APPROVED** [review]
